content/matrices/components/gauss-solver.ts

```diff
@@ -0,0 +1,196 @@
+          inRow1: ${state.inRow1}, inRow2: ${state.inRow2},
+          outRow1: ${state.outRow1}, outRow2: ${state.outRow2}
+        }`);
+      for (const [i, row] of this.$outputCells.entries()) {
+        for (const [j, val] of row.entries()) {
+
+          switch (state.op) {
+            case 'multiply':
+              // "Multiply a row by a non-zero factor"
+              if (state.factor && i === state.outRow1) {
+                const m = Expression.parse(this.$inputCells[state.inRow1][j].text).evaluate();
+                val.text = '' + (state.factor * m);
+              } else {
+                val.text = this.$inputCells[i][j].text;
+              }
+              break;
+            case 'add':
+              // "Add a multiple of one row to another"
+              val.text = this.$inputCells[i][j].text;
+              if (state.factor && i === state.outRow1) {
+                const c1 = Expression.parse(this.$inputCells[state.inRow1][j].text).evaluate();
+                const c2 = Expression.parse(this.$inputCells[state.inRow2][j].text).evaluate();
+                val.text = '' + (c1 + state.factor * c2);
+              } else {
+                val.text = this.$inputCells[i][j].text;
+              }
+              break;
+            case 'swap':
+              // "Swap two rows"
+              if (i === state.outRow1) {
+                val.text = this.$inputCells[state.inRow1][j].text;
+              } else if (i === state.outRow2) {
+                val.text = this.$inputCells[state.inRow2][j].text;
+              } else {
+                val.text = this.$inputCells[i][j].text;
+              }
+              break;
+          }
+        }
+      }
+    });
+
+    // Button to Apply Row Operation
+    const $applyBtn = this.$('.apply');
+    $applyBtn?.on('click', () => {
+      // swapping input and output
+      for (const [i, row] of this.$inputCells.entries()) {
+        for (const [j, val] of row.entries()) {
+          val.text = this.$outputCells[i][j].text;
+        }
+      }
+      this.inputStack.push(this.$outputCells.map(row => row.map(val => val.text)));
+      console.log(this.inputStack);
+
+      if (this.checkForSolvedIdentity()) {
+        console.log('Success!');
+        this.$step.tools.confetti();
+      }
+    });
+
+    const $undoBtn = this.$('.undo');
+    $undoBtn?.on('click', () => {
+      if (this.inputStack.length === 1) return;
+      this.inputStack.pop();
+
+      const newTop = this.inputStack[this.inputStack.length - 1];
+      for (const [i, row] of this.$inputCells.entries()) {
+        for (const [j, val] of row.entries()) {
+          val.text = newTop[i][j];
+        }
+      }
+    });
+
+    // watch for factor change
+    this.model.watch((state) => {
+      let _parseable = true;
+      let expr;
+      let value;
+      try {
+        // FIXME: @philipp this might throw an error, Not sure how else to handle it.
```

Comment:
For now, I would just make factor a number (e.g. `1` or `0.5`). We can allow things like `1/2` in the future, but then we should use a proper equation editor rather than an input field

Reply:
I will try this for now, but I'm not sure it would work with numbers like 1/3 or 1/12.

Reply:
Some things to try:
(a) Rounding? With error margins?
(b) Just switch the problems until we add the expression editor.
(c) Ignore the problem (but China matrix will be close to unsolvable, as checkForSolvedIdentity checks for exact value)